internal/spec/spec.go

```diff
@@ -0,0 +1,143 @@
+//
+//         http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+//    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+//    License for the specific language governing permissions and limitations
+//    under the License.
+
+package spec
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+
+	"github.com/cloudbase/garm-provider-common/cloudconfig"
+	"github.com/cloudbase/garm-provider-common/params"
+	"github.com/cloudbase/garm-provider-common/util"
+	"github.com/cloudbase/garm-provider-oci/config"
+)
+
+func newExtraSpecsFromBootstrapData(data params.BootstrapInstance) (*extraSpecs, error) {
+	spec := &extraSpecs{}
+
+	if len(data.ExtraSpecs) > 0 {
+		if err := json.Unmarshal(data.ExtraSpecs, spec); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal extra specs: %w", err)
+		}
+	}
+
+	return spec, nil
+}
+
+type extraSpecs struct {
+	Ocpus          float32  `json:"ocpus"`
+	MemoryInGBs    float32  `json:"memory_in_gbs"`
+	BootVolumeSize int64    `json:"boot_volume_size"`
+	SSHPublicKeys  []string `json:"ssh_public_keys"`
+}
+
+func GetRunnerSpecFromBootstrapParams(cfg *config.Config, data params.BootstrapInstance, controllerID string) (*RunnerSpec, error) {
+	tools, err := util.GetTools(data.OSType, data.OSArch, data.Tools)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get tools: %s", err)
+	}
+
+	extraSpecs, err := newExtraSpecsFromBootstrapData(data)
+	if err != nil {
+		return nil, fmt.Errorf("error loading extra specs: %w", err)
+	}
+
+	spec := &RunnerSpec{
+		AvailabilityDomain: cfg.AvailabilityDomain,
+		CompartmentID:      cfg.CompartmentId,
+		SubnetID:           cfg.SubnetID,
+		NsgID:              cfg.NsgID,
+		ControllerID:       controllerID,
+		Tools:              tools,
+		BootstrapParams:    data,
+	}
+
+	spec.MergeExtraSpecs(extraSpecs)
+	spec.SetUserData()
+
+	return spec, nil
+}
+
+type RunnerSpec struct {
+	AvailabilityDomain string
+	CompartmentID      string
+	SubnetID           string
+	NsgID              string
+	BootVolumeSize     int64
+	UserData           string
+	ControllerID       string
+	Ocpus              float32
+	MemoryInGBs        float32
+	SSHPublicKeys      []string
+	Tools              params.RunnerApplicationDownload
+	BootstrapParams    params.BootstrapInstance
+}
+
+func (r *RunnerSpec) MergeExtraSpecs(extraSpecs *extraSpecs) {
+	if extraSpecs.Ocpus > 0 {
+		r.Ocpus = extraSpecs.Ocpus
+	} else {
+		r.Ocpus = 1
+	}
+	if extraSpecs.MemoryInGBs > 0 {
+		r.MemoryInGBs = extraSpecs.MemoryInGBs
+	} else {
+		r.MemoryInGBs = 4
+	}
+	if extraSpecs.BootVolumeSize > 0 {
+		r.BootVolumeSize = extraSpecs.BootVolumeSize
+	} else {
+		r.BootVolumeSize = 255
+	}
+	if len(extraSpecs.SSHPublicKeys) > 0 {
+		r.SSHPublicKeys = extraSpecs.SSHPublicKeys
+	}
+}
+
+func (r *RunnerSpec) SetUserData() error {
```

Comment:
Whenever you have a `Set()` function that changes a value in a struct, it's a good idea to put it behind a mutex to prevent race conditions. In that case you'll also need a `Get()` function that also tries to lock the mutex.

```go
func (r *RunnerSpec) SetUserData() error {
   r.mux.Lock()
  defer r.mux.Unlock()
  // rest of code goes here
}
```